internal/service/memorydb/subnet_group.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+	"github.com/hashicorp/terraform-provider-aws/internal/verify"
+)
+
+func ResourceSubnetGroup() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceSubnetGroupCreate,
+		ReadContext:   resourceSubnetGroupRead,
+		UpdateContext: resourceSubnetGroupUpdate,
+		DeleteContext: resourceSubnetGroupDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "Managed by Terraform",
+			},
+			"name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name_prefix"},
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 255),
+					validation.StringDoesNotMatch(
+						regexp.MustCompile(`[-][-]`),
+						"The name may not contain two consecutive hyphens."),
+					validation.StringMatch(
+						// Similar to ElastiCache, MemoryDB normalises names to lowercase.
+						regexp.MustCompile(`^[a-z0-9-]*[a-z0-9]$`),
+						"Only lowercase alphanumeric characters and hyphens allowed. The name may not end with a hyphen."),
+				),
+			},
+			"name_prefix": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name"},
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 255-resource.UniqueIDSuffixLength),
+					validation.StringDoesNotMatch(
+						regexp.MustCompile(`[-][-]`),
+						"The name may not contain two consecutive hyphens."),
+					validation.StringMatch(
+						// Similar to ElastiCache, MemoryDB normalises names to lowercase.
+						regexp.MustCompile(`^[a-z0-9-]+$`),
+						"Only lowercase alphanumeric characters and hyphens allowed."),
+				),
+			},
+			"subnet_ids": {
+				Type:     schema.TypeSet,
+				Required: true,
+				MinItems: 1,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"tags":     tftags.TagsSchema(),
+			"tags_all": tftags.TagsSchemaComputed(),
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+
+		CustomizeDiff: verify.SetTagsDiff,
+	}
+}
+
+func resourceSubnetGroupCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).MemoryDBConn
+	defaultTagsConfig := meta.(*conns.AWSClient).DefaultTagsConfig
+	tags := defaultTagsConfig.MergeTags(tftags.New(d.Get("tags").(map[string]interface{})))
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		name = resource.PrefixedUniqueId(v.(string))
+	} else {
+		name = resource.UniqueId()
+	}
```

Comment:
Drive-by review: You can use the code in the internal `create` package for this.
Documented [here](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/contribution-checklists.md#adding-resource-name-generation-support).

Reply:
Thank you for the hint -- that's now in place.